fog/sample-paykit/src/cached_tx_data/mod.rs

```diff
@@ -390,6 +460,127 @@ impl CachedTxData {
+            if second_block_index - first_block_index > 1 {
+                let block_range = common::BlockRange {
+                    start_block: start_block_index,
+                    end_block: first_block_index + 1,
+                };
+                updated_missed_block_ranges.push(block_range);
+                start_block_index = second_block_index;
+            }
+
+            i += 1;
+            j += 1;
+        }
+
+        let last_block_index = missed_block_indices_sorted.last().unwrap();
+        let last_block_range = common::BlockRange {
+            start_block: start_block_index,
+            end_block: last_block_index + 1,
+        };
+
+        updated_missed_block_ranges.push(last_block_range);
+
+        updated_missed_block_ranges
+    }
+
+    fn create_missed_block_indices(missed_block_ranges: &[common::BlockRange]) -> HashSet<u64> {
+        let mut missed_block_indices: HashSet<u64> = HashSet::new();
+        for missed_block_range in missed_block_ranges {
+            for missed_block_index in missed_block_range.start_block..missed_block_range.end_block {
+                missed_block_indices.insert(missed_block_index);
+            }
+        }
+
+        missed_block_indices
+    }
+
+    // Converts a ledger::BlockResponses to a Vec<TxOutRecord>.
+    fn create_tx_out_records(block_response: &ledger::BlockResponse) -> Vec<TxOutRecord> {
+        let mut tx_out_records: Vec<TxOutRecord> = Vec::new();
+
+        for block_data in &block_response.blocks {
+            let number_of_tx_outs_in_block = block_data.outputs.len() as u64;
+            let first_tx_out_global_index =
+                block_data.global_txo_count - number_of_tx_outs_in_block;
+
+            for (i, external_tx_out) in block_data.outputs.iter().enumerate() {
+                let tx_out: TxOut = TxOut::try_from(external_tx_out).unwrap();
+                let fog_tx_out: FogTxOut = FogTxOut::from(&tx_out);
+
+                let fog_tx_out_metadata: FogTxOutMetadata = FogTxOutMetadata {
+                    global_index: first_tx_out_global_index + i as u64,
+                    block_index: block_data.index,
+                    timestamp: block_data.timestamp,
+                };
+
+                let tx_out_record: TxOutRecord = TxOutRecord::new(fog_tx_out, fog_tx_out_metadata);
+                tx_out_records.push(tx_out_record);
+            }
+        }
+
+        tx_out_records
+    }
+
+    fn convert_block_ranges(
+        common_block_ranges: &[common::BlockRange],
+    ) -> Vec<fog_common::BlockRange> {
+        common_block_ranges
+            .iter()
+            .map(|common_block_range| {
+                let mut fog_common_block_range = fog_common::BlockRange::new();
+                fog_common_block_range.start_block = common_block_range.start_block;
+                fog_common_block_range.end_block = common_block_range.end_block;
+
+                fog_common_block_range
+            })
```

Comment:
I think there is probably a more idiomatic way to do this, like maybe:

* implement a conversion from `common::BlockRange` to `fog_common::BlockRange` (I'm not sure where these two things are coming from, need to check on this)
* When we would call `convert_block_ranges` instead just use the standard library, like, `common_block_ranges.iter().collect::<Vec<fog_common::BlockRange>>()` should work, or maybe it needs `map(BlockRange::from)` also in there

then we would not need this helper function